src/csharp/Grpc.Core/Internal/DefaultChannelCredentialsConfigurator.cs

```diff
@@ -0,0 +1,173 @@
+
+        bool configured;
+        ChannelCredentialsSafeHandle nativeCredentials;
+
+        public ChannelCredentialsSafeHandle NativeCredentials => nativeCredentials;
+        
+        public override void SetInsecureCredentials(object state)
+        {
+            GrpcPreconditions.CheckState(!configured);
+            // null corresponds to insecure credentials.
+            configured = true;
+            nativeCredentials = null;
+        }
+
+        public override void SetSslCredentials(object state, string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback)
+        {
+            GrpcPreconditions.CheckState(!configured);
+            configured = true;
+            nativeCredentials = GetOrCreateNativeCredentials((ChannelCredentials) state,
+                () => CreateNativeSslCredentials(rootCertificates, keyCertificatePair, verifyPeerCallback));
+        }
+
+        public override void SetCompositeCredentials(object state, ChannelCredentials channelCredentials, CallCredentials callCredentials)
+        {
+            GrpcPreconditions.CheckState(!configured);
+            configured = true;
+            nativeCredentials = GetOrCreateNativeCredentials((ChannelCredentials) state,
+                () => CreateNativeCompositeCredentials(channelCredentials, callCredentials));
+        }
+
+        private ChannelCredentialsSafeHandle CreateNativeSslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback)
+        {
+            IntPtr verifyPeerCallbackTag = IntPtr.Zero;
+            if (verifyPeerCallback != null)
+            {
+                verifyPeerCallbackTag = new VerifyPeerCallbackRegistration(verifyPeerCallback).CallbackRegistration.Tag;
+            }
+            return ChannelCredentialsSafeHandle.CreateSslCredentials(rootCertificates, keyCertificatePair, verifyPeerCallbackTag);
+        }
+
+        private ChannelCredentialsSafeHandle CreateNativeCompositeCredentials(ChannelCredentials channelCredentials, CallCredentials callCredentials)
+        {
+            using (var callCreds = callCredentials.ToNativeCredentials())
+            {
+                var nativeComposite = ChannelCredentialsSafeHandle.CreateComposite(channelCredentials.ToNativeCredentials(), callCreds);
+                if (nativeComposite.IsInvalid)
+                {
+                    throw new ArgumentException("Error creating native composite credentials. Likely, this is because you are trying to compose incompatible credentials.");
+                }
+                return nativeComposite;
+            }
+        }
+
+        private ChannelCredentialsSafeHandle GetOrCreateNativeCredentials(ChannelCredentials key, Func<ChannelCredentialsSafeHandle> nativeCredentialsFactory)
+        {
+            Lazy<ChannelCredentialsSafeHandle> lazyValue;
```

Comment:
I think this could be simpler if we just used a mutex here and did a simple cache "lookup or create".... vould then get rid of the while lookup, `Lazy` type, and exception catching.

Personally I wouldn't be concerned about acquiring a global mutex here, since this is only used on the channel construction path (if I'm correct).

Reply:
I thought of using a global lock here, but it seemed more complex than what I have in this PR. I also don't like the idea of holding a global lock when invoking native functions (creating the credentials) and if I wanted to avoid that, it would add even more complexity.

Reply:
I can agree that it's best to avoid calling into c-core with a global lock, and I could see avoiding that while using a mutex would wind up messy.

But different idea: is it possible to remove the global `ConditionalWeakTable` and instead having  each `ChannelCredentials` take ownership of the `ChannelCredentialsConfiguratorBase` object that is used to configured it, i.e. store a `ChannelCredentialsConfiguratorBase` field in each `ChannelCredentials` object.

We could perhaps then continue to just store `Lazy<ChannelCredentialsSafeHandle>` objects, but only now in the `DefaultChannelCredentialsConfigurator` objects - IMO this simplify things compared to the weak map.

Reply:
I think the problem with that approach is:
- you need to make ChannelCredentials (the shared API) remember an instance of channelCredentialsConfiguratorBase just because one of the implementations happens to need it
- the same instance of ChannelCredentials can be used by multiple different configurators (coming from different impls) multiple times. It's not clear what exact configurator should be remembered by the channelCredentials object (the first one ever used?  all of them? - it seems that this can get complex pretty fast).

Reply:
Ok, in keeping the current scheme, I think this loop can still be simplified though.

What do you think about something like this:

```
        private ChannelCredentialsSafeHandle GetOrCreateNativeCredentials(ChannelCredentials key, Func<ChannelCredentialsSafeHandle> nativeCredentialsFactory)
        {
            Lazy<ChannelCredentialsSafeHandle> lazyValue;
            lock (myLock) {
                if (!CachedNativeCredentials.TryGetValue(key, out lazyValue))
                {
                        lazyValue = new Lazy<ChannelCredentialsSafeHandle>(nativeCredentialsFactory);
                        CachedNativeCredentials.Add(key, lazyValue);
                }
            }
            return lazyValue.Value;
        }
```

this would still avoid calling into C-core while holding a c# lock

Reply:
Done.